feat: Upgrade Node Package Manager lock file `package-lock.json` to version 2

dblythy opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Lockfile should be v2 with Node 16

Related issue: #8274
Closes: #8274

### Approach
<!-- Add a description of the approach in this PR. -->

Generates lockfile

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [x] Add tests
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

mtrezza committed:
ci: fix node engine check (#7891)

dblythy committed:
fix: return correct response when revert is used in beforeSave (#7839)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.1 [skip ci]

## [5.2.1-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.0...5.2.1-alpha.1) (2022-03-26)

### Bug Fixes

* return correct response when revert is used in beforeSave ([#7839](https://github.com/parse-community/parse-server/issues/7839)) ([f63fb2b](https://github.com/parse-community/parse-server/commit/f63fb2b338c908f0e7a648d338c26b9daa50c8f2))

dblythy committed:
perf: reduce database operations when using the constant parameter in Cloud Function validation (#7892)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.2 [skip ci]

## [5.2.1-alpha.2](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.1...5.2.1-alpha.2) (2022-03-26)

### Performance Improvements

* reduce database operations when using the constant parameter in Cloud Function validation ([#7892](https://github.com/parse-community/parse-server/issues/7892)) ([48bd512](https://github.com/parse-community/parse-server/commit/48bd512eeb47666967dff8c5e723ddc5b7801daa))

github-actions[bot] committed:
feat: add MongoDB 5.1 compatibility (#7682)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.1 [skip ci]

# [5.3.0-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.2...5.3.0-alpha.1) (2022-03-27)

### Features

* add MongoDB 5.1 compatibility ([#7682](https://github.com/parse-community/parse-server/issues/7682)) ([90155cf](https://github.com/parse-community/parse-server/commit/90155cf1680e5e0499b0000e071c6cb0ce3aef96))

mtrezza committed:
fix: security upgrade parse push adapter from 4.1.0 to 4.1.2 (#7893)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.2 [skip ci]

# [5.3.0-alpha.2](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.1...5.3.0-alpha.2) (2022-03-27)

### Bug Fixes

* security upgrade parse push adapter from 4.1.0 to 4.1.2 ([#7893](https://github.com/parse-community/parse-server/issues/7893)) ([ef56e98](https://github.com/parse-community/parse-server/commit/ef56e98ef65041b4d3b7b82cce3473269c27f6fd))

mtrezza committed:
feat: add MongoDB 5.2 support (#7894)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.3 [skip ci]

# [5.3.0-alpha.3](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.2...5.3.0-alpha.3) (2022-03-27)

### Features

* add MongoDB 5.2 support ([#7894](https://github.com/parse-community/parse-server/issues/7894)) ([6b4b358](https://github.com/parse-community/parse-server/commit/6b4b358f0842ae920e45652f5e8b2afebc6caf3a))

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
Merge branch 'alpha' of https://github.com/dblythy/parse-server into alpha

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
fix: generate lockflie v2

mtrezza commented on .github/workflows/ci.yml:
```diff
@@ -116,7 +116,7 @@ jobs:
         uses: mansona/npm-lockfile-version@v1
         with:
-          version: 1
+          version: 2
```
Good idea to actually keep the version check

dblythy committed:
Delete package-lock.json

dblythy committed:
Create package-lock.json

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
wip

dblythy committed:
Update .eslintrc.json

dblythy committed:
Update package-lock.json

dblythy committed:
Update ci.yml

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

mtrezza commented:
How is this getting along? Seems to be quite an effort looking at the number of commits

mtrezza commented:
@dblythy Did this get stuck for some reason?

dblythy commented:
The main problem is that lockfile v2 is much more strict with dependency mismatches, so there is some refactoring and updates that are required to ensure that `npm install` works. Are we still approaching this update even though we still support Node 14?

mtrezza commented:
We could stay on v1 and exclude this from the initial Parse Server 6 release.

The only argument I see is that the majority of deployments are likely running or will soon run on Node >=16, since Node 14 LTS ends in April. The deployment process will be slower because of the one-time conversion of the lock file. It's also a resource waste. If we take 5.000 Parse Server installations per day (npm stats) and the one-time lock file conversion taking 30 seconds on avg., that would be 42 hours per day or 15,200 hours until Parse Server 7 release in 2024 of wasted compute time.

Lockfile v2 is backwards compatible with v1, but for anyone running custom scripts on the lockfile (like our CI does), this could be a breaking change. But it's a gray zone, we could also merge this anytime throughout the year.

I'll move this from "required" to "optional" in the [tracking list](https://github.com/parse-community/parse-server/issues/8225).

dblythy committed:
changes

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy commented:
I think it's worth migrating to V2 as it also is more restrictive of version conflicts, which can affect Node 14 as well.

dblythy committed:
Update package-lock.json

dblythy committed:
bump runner

dblythy committed:
wip

dblythy committed:
Update package-lock.json

dblythy committed:
Update package-lock.json

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into alpha

dblythy committed:
fix: Nested objects are encoded incorrectly for MongoDB (#8209)

BREAKING CHANGE: Nested objects are now properly stored in the database using JSON serialization; previously, due to a bug only top-level objects were serialized, but nested objects were saved as raw JSON; for example, a nested `Date` object was saved as a JSON object like `{ "__type": "Date", "iso": "2020-01-01T00:00:00.000Z" }` instead of its serialized representation `2020-01-01T00:00:00.000Z` (#8209)

semantic-release-bot committed:
chore(release): 6.0.0-alpha.15 [skip ci]

* Nested objects are encoded incorrectly for MongoDB ([#8209](https://github.com/parse-community/parse-server/issues/8209)) ([1412666](https://github.com/parse-community/parse-server/commit/1412666f75829612de6fb9d7ccae35761c9b75cb))

* Nested objects are now properly stored in the database using JSON serialization; previously, due to a bug only top-level objects were serialized, but nested objects were saved as raw JSON; for example, a nested `Date` object was saved as a JSON object like `{ "__type": "Date", "iso": "2020-01-01T00:00:00.000Z" }` instead of its serialized representation `2020-01-01T00:00:00.000Z` (#8209) ([1412666](1412666))

dblythy committed:
feat: Asynchronous initialization of Parse Server (#8232)

BREAKING CHANGE: This release introduces the asynchronous initialization of Parse Server to prevent mounting Parse Server before being ready to receive request; it changes how Parse Server is imported, initialized and started; it also removes the callback `serverStartComplete`; see the [Parse Server 6 migration guide](https://github.com/parse-community/parse-server/blob/alpha/6.0.0.md) for more details (#8232)

semantic-release-bot committed:
chore(release): 6.0.0-alpha.16 [skip ci]

# [6.0.0-alpha.16](https://github.com/parse-community/parse-server/compare/6.0.0-alpha.15...6.0.0-alpha.16) (2022-12-21)

### Features

* Asynchronous initialization of Parse Server ([#8232](https://github.com/parse-community/parse-server/issues/8232)) ([99fcf45](https://github.com/parse-community/parse-server/commit/99fcf45e55c368de2345b0c4d780e70e0adf0e15))

### BREAKING CHANGES

* This release introduces the asynchronous initialization of Parse Server to prevent mounting Parse Server before being ready to receive request; it changes how Parse Server is imported, initialized and started; it also removes the callback `serverStartComplete`; see the [Parse Server 6 migration guide](https://github.com/parse-community/parse-server/blob/alpha/6.0.0.md) for more details (#8232) ([99fcf45](99fcf45))

dblythy committed:
Update package-lock.json

dblythy committed:
Merge remote-tracking branch 'upstream/alpha' into regenerate-lock

dblythy committed:
Update package-lock.json

dblythy committed:
Update Definitions.js

dblythy committed:
Update package-lock.json

mtrezza commented:
Nice! It seems only deprecation warnings are left now, no more peer dependency warnings.

Rerunning CI

parseplatformorg commented:
🎉 This change has been released in version [6.0.0-alpha.17](https://github.com/parse-community/parse-server/releases/tag/6.0.0-alpha.17)

Moumouls commented:
So nice to be on V2, thanks @dblythy, @mtrezza !

parseplatformorg commented:
🎉 This change has been released in version [6.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/6.0.0-beta.1)

parseplatformorg commented:
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/parse-server/releases/tag/6.0.0)
